File: internal/okf/frontmatter.go
package okf

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"sort"
	"strconv"
	"strings"

	"go.yaml.in/yaml/v3"
)

// The structured face of a document: its frontmatter read as JSON, and
// the same frontmatter written back one key at a time.
//
// This is the surface design doc 0130 §3.2 named as the way back to a
// form editor — "サーバ側に「文書 → 構造」の面を足す", the clause the
// record it folds had written down in advance. What it is not is a second stored form: the
// document stays the concept (0075 §3), this reads one and hands one
// back, and nothing here touches the store.
//
// Why it lives on the server at all, when a form could be filled by a
// parser in the page: the reading half needs YAML, and a hand-written
// subset parser's failure mode is a form that silently drops a producer
// key (0130 §3.1, 0043 §3.6). The writing half needs YAML too, for the
// opposite reason — a browser could emit JSON, which is a subset of
// YAML and therefore always valid, but a `sources:` list flattened onto
// one line is a document nobody wants to read in a diff. Both halves
// are the format's business, and the format's business is here.

// ErrNoFrontmatter is what the structured face answers for text that is
// not an OKF document. Every other operation in this package leaves such
// text alone rather than failing — a move rewrites what it can and stores
// the rest untouched — but there is nothing to project here, and a caller
// filling a form has to be told that rather than shown an empty one.
var ErrNoFrontmatter = errors.New("document has no YAML frontmatter")

// ErrFrontmatterNotAMapping is what it answers for a frontmatter block
// that parses as something other than a mapping: a list, or a bare
// scalar. SPEC §4.1 makes frontmatter a mapping of keys, so this is a
// document no key can be set on.
var ErrFrontmatterNotAMapping = errors.New("frontmatter is not a mapping")

// Structure is one document's frontmatter as structure: the mapping
// decoded to the shapes JSON has, and the key order the document itself
// carries — which JSON objects do not preserve and a form redrawing the
// document would otherwise scramble.
//
// Values are what YAML said, not what ochakai makes of them: no envelope,
// no defaults filled in, no type coerced. A form drawn from this draws
// the document, which is the whole reason the projection (Summary) was
// not widened to do this job instead (0130 §3.1).
type Structure struct {
	Keys   []string       `json:"keys"`
	Values map[string]any `json:"values"`
}

// StructureOf reads a document's frontmatter as structure.
func StructureOf(raw []byte) (*Structure, error) {
	fm, _, ok := splitFrontmatter(string(NormalizeText(raw)))
	if !ok {
		return nil, ErrNoFrontmatter
	}
	// Exactly empty, not merely blank: a frontmatter block holding a
	// lone tab is one YAML refuses to tokenize, and treating it as empty
	// would append a key to a document that does not parse.
	if fm == "" {
		return &Structure{Keys: []string{}, Values: map[string]any{}}, nil
	}
	// mappingKeys, which is the precondition the line surgery rests on:
	// this face reads a document in order to write it back, so a block
	// the removal declines to address is one this must decline to draw
	// as fields. Drawing them and refusing the first save would be a
	// form that can only lie.
	pairs, ok := mappingKeys(fm)
	if !ok {
		return nil, ErrFrontmatterNotAMapping
	}
	out := &Structure{Keys: []string{}, Values: map[string]any{}}
	// keepTimestampText for the reason it exists on the claim path: a
	// date decoded to time.Time has lost the spelling its writer chose,
	// and a form that redraws `stale_after: 2026-12-31` as an instant is
	// one that edited a key nobody touched.
	for _, n := range pairs {
		keepTimestampText(n)
	}
	for i := 0; i+1 < len(pairs); i += 2 {
		key := pairs[i].Value
		var v any
		if err := pairs[i+1].Decode(&v); err != nil {
			continue
		}
		if _, dup := out.Values[key]; !dup {
			out.Keys = append(out.Keys, key)
		}
		out.Values[key] = yamlScalar(v)
	}
	return out, nil
}

// SetFrontmatterKeys returns raw with each key in set given that value
// and each key in unset removed, and every other byte where it was.
//
// A key already in the document is replaced in place — its whole block,
// found by the parse rather than by a scan for "key:" at the start of a
// line, for the reason StripServerKeys reads positions the same way: YAML
// has more than one way to spell a key, and a spelling a scan missed
// would survive and collide with what is appended. A key the document
// does not carry is appended at the end of the frontmatter, which is
// where every key this instance adds goes (appendFrontmatter).
//
// What is not touched: key order, comments, blank lines, the quoting of
// every value the caller did not name, and the body. That is the
// difference between this and rebuilding a document from a form's
// fields, which is the shape that erased data before (documents.js says
// the same thing about the page's own line edits).
//
// A server-owned key is refused rather than written: provenance is what
// this instance observed (design doc 0009), and a face that accepted
// `verified` here would be offering to write it even though the write
// path ignores it — a button that only ever lies.
func SetFrontmatterKeys(raw []byte, set map[string]json.RawMessage, unset []string) ([]byte, error) {
	for _, key := range sortedKeys(set) {
		if serverOwnedKeys[key] {
			return nil, fmt.Errorf("%q is recorded by this instance, not written by hand", key)
		}
	}
	for _, key := range unset {
		if serverOwnedKeys[key] {
			return nil, fmt.Errorf("%q is recorded by this instance, not written by hand", key)
		}
	}
	doc := NormalizeText(raw)
	fm, rest, ok := splitFrontmatter(string(doc))
	if !ok {
		return nil, ErrNoFrontmatter
	}
	present, err := topLevelKeys(fm)
	if err != nil {
		return nil, err
	}

	// Replace and remove in one pass over the lines, so the ranges the
	// parse produced all refer to the same text. Doing them one key at a
	// time would need a reparse between each, and a caller setting two
	// keys would pay for a document rewritten twice.
	replace := map[string]string{}
	for key, v := range set {
		block, err := renderKey(key, v)
		if err != nil {
			return nil, err
		}
		replace[key] = block
	}
	remove := map[string]bool{}
	for _, key := range unset {
		remove[key] = true
	}

	// A plain split is exact here because mappingKeys has already
	// refused every block whose lines the parser would count differently
	// — a bare carriage return, the Unicode breaks, a document marker.
	// An empty block has no lines at all: splitting "" would yield one
	// empty string, a blank line the writer never wrote.
	var lines []string
	if fm != "" {
		lines = strings.Split(strings.TrimSuffix(fm, "\n"), "\n")
	}
	// 1-based, as YAML counts, and one longer so the last line's index
	// is addressable.
	act := make([]string, len(lines)+2)
	drop := make([]bool, len(lines)+2)
	for _, key := range present {
		_, replacing := replace[key]
		if !replacing && !remove[key] {
			continue
		}
		// Every occurrence goes, and the new block takes the place of
		// the first. A document naming one key twice is one yaml.v3
		// parses and this could otherwise leave half of — a second
		// `title:` under a form that reports it set the title once.
		for i, r := range keyRange(fm, len(lines), key) {
			if i == 0 && replacing {
				act[r.from] = key
			}
			for line := r.from; line <= r.to; line++ {
				drop[line] = true
			}
		}
	}

	var kept []string
	for i := range lines {
		if key := act[i+1]; key != "" {
			kept = append(kept, strings.Split(strings.TrimSuffix(replace[key], "\n"), "\n")...)
			delete(replace, key)
		}
		if drop[i+1] {
			continue
		}
		kept = append(kept, lines[i])
	}
	// Whatever is left in replace is a key the document did not carry.
	// Appended in a stable order, because two runs of the same request
	// must produce the same bytes — the content hash is taken over them.
	for _, key := range sortedKeys(replace) {
		kept = append(kept, strings.Split(strings.TrimSuffix(replace[key], "\n"), "\n")...)
	}
	out := []byte("---\n---\n" + rest)
	if len(kept) > 0 {
		out = []byte("---\n" + strings.Join(kept, "\n") + "\n---\n" + rest)
	}
	// Normalized on the way out as well as on the way in, so that setting
	// the same key twice produces the same bytes: what a caller stores
	// has to be what a caller storing it again would store. One pass is
	// one, since NormalizeText is idempotent.
	out = NormalizeText(out)
	if err := verifySpliced(out, set, unset); err != nil {
		return nil, err
	}
	return out, nil
}

// verifySpliced reads the produced document back and requires it to say
// what the caller asked for: every key set is there, every key unset is
// gone. Presence, not value — YAML retypes a scalar on the way through
// and that is the format's business, not a failure.
//
// It is here because the splice is line arithmetic over text a producer
// wrote, and a frontmatter block can be shaped so that the arithmetic is
// right and the result still does not read back — a line of whitespace
// that swallows what follows it, an anchor, a spelling the parser and
// the line map disagree about. Rather than enumerate those, the function
// checks the one thing that matters: **a document handed back is one the
// next read agrees with.** A form whose save silently did nothing, and
// whose next save then wrote the key twice, is the failure mode design
// doc 0130 §3.1 refused a browser-side parser over — it must not come back
// on this side either.
func verifySpliced(out []byte, set map[string]json.RawMessage, unset []string) error {
	if len(set) == 0 && len(unset) == 0 {
		return nil
	}
	keys, err := topLevelKeys(mustFrontmatter(out))
	if err != nil {
		return fmt.Errorf("the result would not parse: %w", err)
	}
	has := make(map[string]bool, len(keys))
	for _, k := range keys {
		has[k] = true
	}
	for _, k := range sortedKeys(set) {
		if !has[k] {
			return fmt.Errorf("this frontmatter cannot be edited by key: %q did not read back after it was written", k)
		}
	}
	for _, k := range unset {
		if has[k] {
			return fmt.Errorf("this frontmatter cannot be edited by key: %q is still there after it was removed", k)
		}
	}
	return nil
}

func mustFrontmatter(doc []byte) string {
	fm, _, _ := splitFrontmatter(string(doc))
	return fm
}

// keyRange is ownedKeyRanges for one named key — every line range the
// key occupies, in document order.
func keyRange(fm string, lines int, key string) []lineRange {
	return ownedKeyRanges(fm, lines, func(k string) bool { return k == key })
}

// topLevelKeys names the frontmatter's own keys, in document order, or
// reports that the block is not one this face can address (mappingKeys).
func topLevelKeys(fm string) ([]string, error) {
	if fm == "" {
		return nil, nil
	}
	pairs, ok := mappingKeys(fm)
	if !ok {
		return nil, ErrFrontmatterNotAMapping
	}
	out := make([]string, 0, len(pairs)/2)
	for i := 0; i+1 < len(pairs); i += 2 {
		out = append(out, pairs[i].Value)
	}
	return out, nil
}

// renderKey renders one key and its value as the YAML lines that stand
// for it — block style, indented as the rest of the frontmatter is, so
// what a form writes reads like what a person writes.
//
// The value arrives as JSON rather than as a decoded map because **JSON
// objects are ordered on the wire and Go maps are not**: yaml.Marshal of
// a map sorts, so a source written from a form came out
// `author, id, last_modified, resource, …` while the bundle writer emits
// SPEC's own order. Walking the JSON keeps whatever order the caller
// wrote, which is the caller's business and not this package's opinion —
// the same posture every other edit here takes toward a document.
func renderKey(key string, v json.RawMessage) (string, error) {
	if key == "" || strings.ContainsAny(key, ":\n#") {
		return "", fmt.Errorf("%q is not a frontmatter key", key)
	}
	value, err := jsonAsYAML(v)
	if err != nil {
		return "", fmt.Errorf("%q: %w", key, err)
	}
	// yaml.Marshal, with no indent set, so a block written from a field
	// is indented exactly as the bundle writer (Canonical) would have
	// written it. A document whose sources block is indented differently
	// from the one beside it is a diff nobody asked for, and the one
	// spelling that cannot be wrong here is this package's own.
	out, err := yaml.Marshal(&yaml.Node{
		Kind: yaml.MappingNode,
		Content: []*yaml.Node{
			{Kind: yaml.ScalarNode, Tag: "!!str", Value: key}, value,
		},
	})
	if err != nil {
		return "", err
	}
	// The key has to survive its own rendering, which is not something
	// the type system says: a key whose bytes are not valid UTF-8 goes
	// out as `!!binary qA==`, comes back as something else, and is then
	// appended a second time by the next write because the first is no
	// longer found. Asking the parser closes that whole class rather
	// than the one spelling the fuzzer happened to reach.
	if back, err := topLevelKeys(string(out)); err != nil || len(back) != 1 || back[0] != key {
		return "", fmt.Errorf("%q is not a frontmatter key ochakai can write", key)
	}
	return string(out), nil
}

func sortedKeys[V any](m map[string]V) []string {
	out := make([]string, 0, len(m))
	for k := range m {
		out = append(out, k)
	}
	sort.Strings(out)
	return out
}

// jsonAsYAML turns one JSON value into the YAML node that stands for it,
// in the order the JSON was written.
//
// Strings carry an explicit !!str tag so a value that looks like a date
// or a number keeps its quotes and comes back the string it went in as
// (design doc 0036 §3.7 is the same rule for stale_after), and the one
// shape yaml would emit as a block scalar the parser cannot read back is
// double-quoted instead — which is what the `text` type does for every
// other value this package writes.
func jsonAsYAML(raw json.RawMessage) (*yaml.Node, error) {
	dec := json.NewDecoder(bytes.NewReader(raw))
	dec.UseNumber()
	n, err := jsonNode(dec)
	if err != nil {
		return nil, err
	}
	if _, err := dec.Token(); !errors.Is(err, io.EOF) {
		return nil, errors.New("a value must be a single JSON value")
	}
	return n, nil
}

func jsonNode(dec *json.Decoder) (*yaml.Node, error) {
	tok, err := dec.Token()
	if err != nil {
		return nil, err
	}
	switch t := tok.(type) {
	case json.Delim:
		switch t {
		case '[':
			seq := &yaml.Node{Kind: yaml.SequenceNode, Tag: "!!seq"}
			for dec.More() {
				item, err := jsonNode(dec)
				if err != nil {
					return nil, err
				}
				seq.Content = append(seq.Content, item)
			}
			_, err := dec.Token() // ]
			return seq, err
		case '{':
			m := &yaml.Node{Kind: yaml.MappingNode, Tag: "!!map"}
			for dec.More() {
				name, err := dec.Token()
				if err != nil {
					return nil, err
				}
				key, ok := name.(string)
				if !ok {
					return nil, errors.New("a mapping key must be a string")
				}
				value, err := jsonNode(dec)
				if err != nil {
					return nil, err
				}
				m.Content = append(m.Content,
					&yaml.Node{Kind: yaml.ScalarNode, Tag: "!!str", Value: key}, value)
			}
			_, err := dec.Token() // }
			return m, err
		}
		return nil, errors.New("unbalanced JSON")
	case string:
		if blockScalarRisk(t) {
			return &yaml.Node{Kind: yaml.ScalarNode, Tag: "!!str", Style: yaml.DoubleQuotedStyle, Value: t}, nil
		}
		return &yaml.Node{Kind: yaml.ScalarNode, Tag: "!!str", Value: t}, nil
	case json.Number:
		tag := "!!int"
		if strings.ContainsAny(t.String(), ".eE") {
			tag = "!!float"
		}
		return &yaml.Node{Kind: yaml.ScalarNode, Tag: tag, Value: t.String()}, nil
	case bool:
		return &yaml.Node{Kind: yaml.ScalarNode, Tag: "!!bool", Value: strconv.FormatBool(t)}, nil
	case nil:
		// A key set to null is a key that says nothing. Written rather
		// than refused: the document is the writer's, and `null` is a
		// value YAML has.
		return &yaml.Node{Kind: yaml.ScalarNode, Tag: "!!null", Value: "null"}, nil
	}
	return nil, fmt.Errorf("unsupported JSON value %T", tok)
}

File: internal/webui/static/js/frontmatter.js
// The frontmatter as fields: which keys get an editor, what each one
// looks like, and how what a person typed becomes the value that goes
// back into the document.
//
// Pure, like documents.js and for the same reason — this decides what a
// writer's frontmatter becomes, so it is held to examples. It renders
// markup from values and reads values back out of plain objects the
// editor collected from the DOM; the DOM itself is views/editor.js.
//
// It parses no YAML and writes none. Both directions go through
// POST /api/v1/frontmatter (design doc 0130): the server reads the
// document a producer wrote and renders the block a person will read in
// a diff, because a subset parser in the page is the thing design doc
// 0130 §3.1 refused — its failure mode is a form that silently drops a
// key.

import { esc } from './escape.js';
import { KNOWN_TYPES, STATUSES } from './vocab.js';

// The runtimes SPEC §10.2 names as examples. A suggestion list rather
// than a vocabulary: the key takes any string, and ochakai stores what
// the writer wrote (design doc 0036 §3.4).
const RUNTIMES = ['bigquery', 'postgres', 'dbt', 'python', 'looker'];

// FIELDS is every frontmatter key this form edits, in the order OKF
// SPEC introduces them — base (§4.1), provenance (§5.1), lifecycle
// (§5.4-§5.5), Attested Computation (§10.2) — which is the order the
// bundle writer emits them in too, so a document written here reads like
// one written by hand.
//
// What is not here is what nobody writes: `generated` and `verified` are
// this instance's observations and the write path ignores them (design
// doc 0009), and `okf_version` belongs to the bundle root, which is
// generated. A producer's own key is not here either — it has no field,
// and it is not lost: the document keeps it, and the form says so.
export const FIELDS = [
  { key: 'type', kind: 'combo', label: '型', options: KNOWN_TYPES, required: true },
  { key: 'title', kind: 'text', label: 'タイトル' },
  { key: 'description', kind: 'prose', label: '説明', hint: '一文で。検索結果と一覧に出ます。' },
  {
    key: 'resource', kind: 'text', label: '実体の場所',
    placeholder: 'bigquery://project/dataset/table',
    hint: 'この知識が指している物の URI。テーブル、ダッシュボード、外部ページなど。',
  },
  { key: 'tags', kind: 'list', label: 'タグ', placeholder: 'sales' },
  {
    key: 'sources', kind: 'rows', label: '出典', addLabel: '出典を足す',
    hint: 'この知識が何から導かれたかです。',
    columns: [
      { name: 'resource', label: '場所', placeholder: 'bigquery://project/dataset/table' },
      { name: 'title', label: '見出し' },
      { name: 'id', label: '識別子' },
      { name: 'author', label: '作成者', placeholder: 'human:tanaka@example.co.jp' },
      { name: 'usage_count', label: '参照回数', kind: 'number' },
      { name: 'last_modified', label: '最終更新', kind: 'date' },
    ],
  },
  {
    key: 'usage_window', kind: 'object', label: '参照回数を数えた期間', onlyIfPresent: true,
    hint: '上の参照回数が、いつからいつまでを数えたものかです。',
    columns: [
      { name: 'from', label: '開始', kind: 'date' },
      { name: 'to', label: '終了', kind: 'date' },
    ],
  },
  { key: 'status', kind: 'enum', label: '状態', options: STATUSES },
  { key: 'stale_after', kind: 'date', label: '再確認日' },
  {
    key: 'runtime', kind: 'combo', label: '実行環境', options: RUNTIMES,
    types: ['Attested Computation'],
    hint: '計算を約束する型では必須です(SPEC §10.2)。ochakai は記録するだけで、実行はしません。',
  },
  {
    key: 'parameters', kind: 'rows', label: 'パラメータ', addLabel: 'パラメータを足す',
    types: ['Attested Computation'],
    hint: 'エージェントが埋める穴です。',
    columns: [
      { name: 'name', label: '名前', placeholder: 'year' },
      { name: 'type', label: '型', placeholder: 'integer' },
      { name: 'required', label: '必須', kind: 'bool' },
    ],
  },
  {
    key: 'computation', kind: 'text', label: '計算の置き場所',
    types: ['Attested Computation'],
    hint: '空のままなら、本文の <code># Computation</code> のコードブロックが計算です。',
  },
  {
    key: 'executor', kind: 'object', label: '実行のしかた',
    types: ['Attested Computation'],
    columns: [
      { name: 'resource', label: '手順・コードの場所' },
      { name: 'receipt', label: '実行が返すべき項目', kind: 'list' },
    ],
  },
  {
    key: 'attester', kind: 'object', label: '検証のしかた',
    types: ['Attested Computation'],
    hint: '実行が正しく行われたかを機械的に確かめるコードの場所です。',
    columns: [{ name: 'resource', label: 'コードの場所' }],
  },
];

export const FIELD_KEYS = new Set(FIELDS.map(f => f.key));

// otherKeys names the frontmatter keys the form has no editor for, in
// the document's own order. They are the reason this page can be a form
// at all: a producer's key is kept exactly where its writer put it (SPEC
// §4.1), and saying so is what keeps "the form is the document" true
// rather than merely hoped for.
export function otherKeys(keys) {
  return (keys || []).filter(k => !FIELD_KEYS.has(k));
}

// ---- values -----------------------------------------------------------

// valueOf turns what the editor collected for one field into the value
// that goes into the document, or undefined for "this key is not in the
// document" — which is not the same as an empty one. ochakai does not
// write a key its writer left out (design doc 0046 §3.9), so a cleared
// field removes the key rather than storing "".
export function valueOf(field, raw) {
  switch (field.kind) {
    case 'list':
      return listValue(raw);
    case 'rows': {
      const rows = (raw || []).map(r => objectValue(field.columns, r)).filter(Boolean);
      return rows.length ? rows : undefined;
    }
    case 'object':
      return objectValue(field.columns, raw);
    default:
      return scalarValue({ kind: field.kind }, raw);
  }
}

function listValue(raw) {
  const out = (raw || []).map(v => String(v ?? '').trim()).filter(v => v !== '');
  return out.length ? out : undefined;
}

// objectValue drops every column left empty and the object itself when
// nothing is left — a `usage_window` with neither end is a key that says
// nothing, and writing one would be inventing a claim.
function objectValue(columns, raw) {
  const out = {};
  for (const col of columns) {
    const v = col.kind === 'list' ? listValue((raw || {})[col.name]) : scalarValue(col, (raw || {})[col.name]);
    if (v !== undefined) out[col.name] = v;
  }
  return Object.keys(out).length ? out : undefined;
}

function scalarValue(col, raw) {
  if (col.kind === 'bool') {
    // SPEC §10.2: absent and false say the same thing, so only true is
    // written. A `required: false` line would be a key the writer did
    // not ask for.
    return raw === true ? true : undefined;
  }
  const s = String(raw ?? '').trim();
  if (s === '') return undefined;
  if (col.kind === 'number') {
    const n = Number(s);
    // Not a number is kept as the text it was: the server stores what
    // the writer wrote, and a form that silently turned "many" into NaN
    // would be deciding for them.
    return Number.isFinite(n) ? n : s;
  }
  return s;
}

// ---- markup -----------------------------------------------------------

// visibleFields picks the fields worth drawing for this document. A key
// the document carries always gets its editor — hiding one would make it
// uneditable without saying so. A field tied to types (the Attested
// Computation contract, SPEC §10.2) appears only when the type asks for
// it, and one marked onlyIfPresent stays away until a document brings
// the key: every field at once was a form nobody could see whole.
export function visibleFields(values) {
  const v = values || {};
  return FIELDS.filter(f => {
    if (v[f.key] !== undefined) return true;
    if (f.types) return f.types.includes(v.type);
    return !f.onlyIfPresent;
  });
}

// fieldsMarkup renders the visible fields from the frontmatter's values.
// The values come from the server's read of the document, so what is
// drawn here is the document — not a second store the page keeps beside
// it.
export function fieldsMarkup(values) {
  const v = values || {};
  return visibleFields(v).map(f => fieldMarkup(f, v[f.key])).join('');
}

export function fieldMarkup(field, value) {
  const body = {
    list: listMarkup, rows: rowsMarkup, object: objectMarkup,
  }[field.kind];
  if (body) return groupMarkup(field, body(field, value));
  return `
      <div class="field" data-fm-field="${field.key}">
        ${labelMarkup(field, `f-${field.key}`)}
        ${inputMarkup(field, `f-${field.key}`, field.key, '', value)}
        ${hintMarkup(field)}
      </div>`;
}

function groupMarkup(field, inner) {
  return `
      <div class="fieldset" data-fm-field="${field.key}">
        <div class="legend">${esc(field.label)} <code>${field.key}</code>${field.required ? ' <span class="req">必須</span>' : ''}</div>
        ${inner}
        ${hintMarkup(field)}
      </div>`;
}

function labelMarkup(field, id) {
  return `<label class="top" for="${id}">${esc(field.label)} <code>${field.key}</code>`
    + `${field.required ? ' <span class="req">必須</span>' : ''}</label>`;
}

function hintMarkup(field) {
  // The hint is authored copy, not a value, so its markup is deliberate:
  // it is the one string here that is not escaped.
  return field.hint ? `<div class="hint">${field.hint}</div>` : '';
}

// inputMarkup is one control. name and col address it: the editor reads
// the form back by walking [data-fm-key], so every control says which
// key it belongs to, which row it is in, and which column of that row.
function inputMarkup(field, id, key, path, value, col) {
  const spec = col || field;
  const at = `data-fm-key="${key}"${path ? ` data-fm-path="${path}"` : ''}`;
  const place = spec.placeholder ? ` placeholder="${esc(spec.placeholder)}"` : '';
  switch (spec.kind) {
    case 'prose':
      return `<textarea id="${id}" ${at} rows="2"${place}>${esc(value ?? '')}</textarea>`;
    case 'bool':
      // A bare checkbox under the column's own label, like every other
      // control here: a checkbox carrying its own text sat a line above
      // the boxes beside it, and lining it up by hand is a number that
      // goes wrong the first time a font changes.
      return `<input type="checkbox" class="check" id="${id}" ${at}${value === true ? ' checked' : ''}>`;
    case 'date':
      // A date input takes YYYY-MM-DD and hands one back, which is what
      // OKF writes (SPEC §5.5). A value in any other spelling would be
      // dropped by the control, so it falls back to a text box rather
      // than being erased by being shown.
      return isDate(value) || !value
        ? `<input type="date" id="${id}" ${at} value="${esc(value ?? '')}">`
        : `<input type="text" id="${id}" ${at} class="mono" value="${esc(value)}">`;
    case 'number':
      return `<input type="number" id="${id}" ${at} value="${esc(value ?? '')}"${place}>`;
    case 'enum':
      return `<select id="${id}" ${at}>`
        + `<option value=""${value ? '' : ' selected'}>—</option>`
        + spec.options.map(o => `<option value="${esc(o)}"${o === value ? ' selected' : ''}>${esc(o)}</option>`).join('')
        + `</select>`;
    case 'combo':
      // A list of suggestions, not a vocabulary: the key takes any
      // string, and a <select> would make the page refuse what the
      // format allows.
      return `<input type="text" id="${id}" ${at} list="fm-list-${key}" value="${esc(value ?? '')}"${place}>`
        + `<datalist id="fm-list-${key}">${spec.options.map(o => `<option value="${esc(o)}"></option>`).join('')}</datalist>`;
    default:
      return `<input type="text" id="${id}" ${at} value="${esc(value ?? '')}"${place}>`;
  }
}

const isDate = v => typeof v === 'string' && /^\d{4}-\d{2}-\d{2}$/.test(v);

function listMarkup(field, value) {
  const items = Array.isArray(value) ? value : [];
  const rows = items.map((v, i) => listItemMarkup(field.key, '', i, v, field)).join('');
  return `<div class="fm-list" data-fm-list="${field.key}">${rows}</div>`
    + `<button type="button" class="btn small fm-add" data-fm-add="${field.key}">＋</button>`;
}

// listItemMarkup is one string of a list of strings, with the button
// that takes it out. Exported so the editor can add a row without
// redrawing the form under a cursor that is in it.
export function listItemMarkup(key, path, i, value, field) {
  const place = field && field.placeholder ? ` placeholder="${esc(field.placeholder)}"` : '';
  return `<div class="fm-item">`
    + `<input type="text" data-fm-key="${key}"${path ? ` data-fm-path="${path}"` : ''} data-fm-i="${i}" value="${esc(value ?? '')}"${place}>`
    + `<button type="button" class="btn small fm-del" title="この行を消します" aria-label="この行を消す">×</button>`
    + `</div>`;
}

function objectMarkup(field, value) {
  const v = value && typeof value === 'object' && !Array.isArray(value) ? value : {};
  return `<div class="row-trio">` + field.columns.map(col => {
    const id = `f-${field.key}-${col.name}`;
    if (col.kind === 'list') {
      const items = Array.isArray(v[col.name]) ? v[col.name] : [];
      return `<div class="field"><label class="top">${esc(col.label)} <code>${col.name}</code></label>`
        + `<div class="fm-list" data-fm-list="${field.key}" data-fm-list-path="${col.name}">`
        + items.map((s, i) => listItemMarkup(field.key, col.name, i, s)).join('')
        + `</div><button type="button" class="btn small fm-add" data-fm-add="${field.key}" data-fm-add-path="${col.name}">＋</button></div>`;
    }
    return `<div class="field"><label class="top" for="${id}">${esc(col.label)} <code>${col.name}</code></label>`
      + inputMarkup(field, id, field.key, col.name, v[col.name], col)
      + (col.hint ? `<div class="hint">${col.hint}</div>` : '')
      + `</div>`;
  }).join('') + `</div>`;
}

function rowsMarkup(field, value) {
  const rows = Array.isArray(value) ? value : [];
  return `<div class="rows" data-fm-rows="${field.key}">${rows.map((r, i) => rowMarkup(field, i, r)).join('')}</div>`
    + `<button type="button" class="btn small fm-add" data-fm-add="${field.key}">${esc(field.addLabel || '行を足す')}</button>`;
}

// rowMarkup is one repeated object — one source, one parameter. Exported
// for the same reason listItemMarkup is.
export function rowMarkup(field, i, row) {
  const r = row && typeof row === 'object' ? row : {};
  const cells = field.columns.map(col => {
    const id = `f-${field.key}-${i}-${col.name}`;
    return `<div class="field">`
      + `<label class="top" for="${id}">${esc(col.label)} <code>${col.name}</code></label>`
      + inputMarkup(field, id, field.key, col.name, r[col.name], { ...col, kind: col.kind || 'text' })
      + (col.hint ? `<div class="hint">${col.hint}</div>` : '')
      + `</div>`;
  }).join('');
  return `<div class="row-card" data-fm-i="${i}">`
    + `<div class="row-head"><span class="n">${i + 1}</span>`
    + `<button type="button" class="btn small fm-del" title="この行を消します" aria-label="この行を消す">×</button></div>`
    + `<div class="row-trio">${cells}</div></div>`;
}
